### PR TITLE
Add an Inductor decomposition for aten.polar, pytorch/pytorch #192315 (#192315)

### PR DESCRIPTION
**Pull request by simranbarnwal**
Summary:

Decompose aten.polar in Inductor into real ops: compute abs*cos(angle) and
abs*sin(angle), interleave them into the last dim, and reinterpret as complex via
aten.view.dtype (mirroring the existing complex `add` decomposition). This lets
backends without complex codegen (e.g. Triton) lower polar.

Test Plan:
test_polar in caffe2/test/inductor/test_torchinductor.py (compiles torch.polar and
checks it matches eager on CPU and GPU).

Reviewed By: nandesuka

@diff-train-skip-merge

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -8197,6 +8197,27 @@ def fn(dist, angle):
         )
         self.common(fn, (*inp,), reference_in_float=self.device != "mps")
 
+    def test_polar_empty(self):
```
Comment (blaine-rister) [inline test/inductor/test_torchinductor.py]: nit: "empty" makes me think of `torch.empty`, i.e. uninitialized. Perhaps this test refers to size-0 tensors? It might be worthwhile to clarify the terminology.

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py [hunk 2]
```diff
@@ -341,6 +341,15 @@ def run(*ex, **kwargs):
     "test_pointwise_laguerre_polynomial_l_dynamic_shapes": TestFailure(("cuda", "xpu")),
     "test_pointwise_legendre_polynomial_p_dynamic_shapes": TestFailure(("cuda", "xpu")),
     "test_polar_dynamic_shapes": TestFailure(("cpu", "cuda", "xpu"), is_skip=True),
+    # Degenerate shapes: an empty or 0-dim input produces no dynamic loop, so
```
Comment (blaine-rister) [inline test/inductor/test_torchinductor_codegen_dynamic_shapes.py]: if the reasons are the same as above, do we need a separate comment here?

File: test/inductor/test_torchinductor_opinfo.py [hunk 3]
```diff
@@ -251,6 +251,12 @@ def format_op(op):
     "resize_": {b8, f16, f32, f64, i32, i64},
     "resize_as_": {b8, f16, f32, f64, i32, i64},
     "histc": {f16},
+    # Same reason as "complex"/"view_as_complex" above: check_model runs the
```
Comment (blaine-rister) [inline test/inductor/test_torchinductor_opinfo.py]: same comment as the previous test failures

File: test/test_tensor_creation_ops.py [hunk 4]
```diff
@@ -447,17 +447,19 @@ def test_torch_complex(self, device, dtype):
...
+    @dtypes(torch.half, torch.float32, torch.float64)
     def test_torch_polar(self, device, dtype):
         abs = torch.tensor([1, 2, -3, -4.5, 1, 1], device=device, dtype=dtype)
         angle = torch.tensor([math.pi / 2, 5 * math.pi / 4, 0, -11 * math.pi / 6, math.pi, -math.pi],
                              device=device, dtype=dtype)
         z = torch.polar(abs, angle)
-        complex_dtype = torch.complex64 if dtype == torch.float32 else torch.complex128
+        complex_dtype = float_to_corresponding_complex_type_map[dtype]
+        # float16 loses precision in cos/sin and the interleaved storage; relax tol.
+        atol, rtol = (1e-2, 1e-2) if dtype == torch.half else (1e-5, 1e-5)
```
Comment (blaine-rister) [inline test/test_tensor_creation_ops.py]: very nit: consider parametrizing on `(dtype, tol)` as a pair these in the decorator. That may improve clarity vs. deciding the tolerance dynamically.

File: torch/_inductor/decomposition.py [hunk 5]
```diff
@@ -699,6 +699,35 @@ def reshape_tensor_complex(tensor: torch.Tensor) -> torch.Tensor:
...
+    # and squeeze the result back to 0-dim via output_size_zero.
+    output_size_zero = abs.ndim == 0 and angle.ndim == 0
+    if abs.ndim == 0:
+        abs = abs.reshape(1)
+    if angle.ndim == 0:
+        angle = angle.reshape(1)
+    re = abs * torch.cos(angle)
+    im = abs * torch.sin(angle)
+    if re.numel() == 0:
+        # aten.view.dtype requires stride(-1) == 1, but Inductor collapses the
```
Comment (blaine-rister) [inline torch/_inductor/decomposition.py]: nit: This comment feels pretty verbose and specific to the current state of the codebase. It might go out of date over time. Can we summarize it? Would it suffice to say there is nothing to compute when the tensor's size is 0?

File: torch/_inductor/decomposition.py [hunk 6]
```diff
@@ -699,6 +699,35 @@ def reshape_tensor_complex(tensor: torch.Tensor) -> torch.Tensor:
     return result
 
 
+@register_decomposition([aten.polar])
+def polar(abs: torch.Tensor, angle: torch.Tensor) -> torch.Tensor:
+    complex_type = utils.corresponding_complex_dtype(abs.dtype)
+    # flatten(start_dim=-2) below needs >=2 dims, so promote 0-dim inputs to [1]
+    # and squeeze the result back to 0-dim via output_size_zero.
+    output_size_zero = abs.ndim == 0 and angle.ndim == 0
```
Comment (blaine-rister) [inline torch/_inductor/decomposition.py]: nit: what about storing the original output size, then calling `Tensor.reshape` to return to that? This might be a bit more generic or easy to understand than a special case for rank-0.

File: test/inductor/test_torchinductor.py [hunk 7]
```diff
@@ -8197,6 +8197,27 @@ def fn(dist, angle):
...
+        # requires stride(-1) == 1.
+        def fn(dist, angle):
+            return torch.polar(dist, angle)
+
+        dtype = highest_precision_float(self.device)
+        for shape in [(0,), (0, 10, 3), (2, 0, 3)]:
+            inp = (torch.rand(shape, dtype=dtype), torch.rand(shape, dtype=dtype))
+            self.common(fn, inp, reference_in_float=self.device != "mps")
+
+    def test_polar_scalar(self):
```
Comment (blaine-rister) [inline test/inductor/test_torchinductor.py:8209]: Can we also test fusion of polar with elementwise and/or reduction operators? That might help to guard against any bugs in the view/reshapelogic.

**Review by blaine-rister — COMMENTED** [review]
This mostly LGTM, but I left a few comments, mostly about testing.

**Comment by blaine-rister** [comment]
Do we have any performance data comparing the Inductor decomp/lowering to the existing ATen fallback on GPU? I'm wondering if we could run a microbenchmark on A100 or H100 just to sanity check that there's no large regression.

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (simranbarnwal) [reply test/inductor/test_torchinductor.py]: renamed to `test_polar_size_zero`, which is what it actually covers.

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py (hunk 2 above)
Comment (simranbarnwal) [reply test/inductor/test_torchinductor_codegen_dynamic_shapes.py]: Removed; it's the same reason as the neighbouring entries.

File: test/inductor/test_torchinductor_opinfo.py (hunk 3 above)
Comment (simranbarnwal) [reply test/inductor/test_torchinductor_opinfo.py]: Agreed, that landed before I got to it. Trimming it.

File: test/test_tensor_creation_ops.py (hunk 4 above)
Comment (simranbarnwal) [reply test/test_tensor_creation_ops.py]: Tried it, but `@parametrize("dtype,tol", ...)` collides with the device-type suffix and produces `test_torch_polar_float32_tol_1e-05_cpu_float32` — dtype twice, and it breaks any skip-list entry naming `test_torch_polar_cpu_float16`. Following up with `@dtypes` plus an explicit `tol = {torch.half: 1e-2}.get(dtype, 1e-5)`; happy to switch if you'd rather have the parametrized names.

File: torch/_inductor/decomposition.py (hunk 5 above)
Comment (simranbarnwal) [reply torch/_inductor/decomposition.py]: Trimmed to two lines: nothing to compute, and `aten.view.dtype` can't reinterpret a 0-element tensor.

File: torch/_inductor/decomposition.py (hunk 6 above)
Comment (simranbarnwal) [reply torch/_inductor/decomposition.py]: Done, dropped `output_size_zero` for `out_shape = re.shape` + flatten/reshape. Verified against 0-dim, 1–5d, size-0, broadcasting and dynamic shapes.

File: test/inductor/test_torchinductor.py (hunk 7 above)
Comment (simranbarnwal) [reply test/inductor/test_torchinductor.py:8209]: Added `test_polar_fused`: `polar(dist*2, angle+0.1)` feeding `abs().sum()`, `.real*3` and `.imag.amax()`.

**Comment by simranbarnwal** [comment]
> Do we have any performance data comparing the Inductor decomp/lowering to the existing ATen fallback on GPU? I'm wondering if we could run a microbenchmark on A100 or H100 just to sanity check that there's no large regression.

Benchmarked on A100 (`do_bench`, GPU time): parity at large sizes: 4096² 1.04x/1.00x and 8192² 1.03x/1.00x (f32/f64), both at ~1.35 TB/s, but a real regression at small/mid: 64x1024 is 0.44x f32 / 0.29x f64, 1024x1024 0.66x f32. 

Cause is in the codegen: `stack + flatten` gives one thread per scalar lane rather than per complex element (`xnumel` = 2N, `x0 = xindex % 2`), each input loaded twice under complementary masks, and both `cos` and `sin` evaluated per thread with a `tl.where` to select. That's inherent to the interleave-via-stack pattern inherited from the complex `add` decomp, worth a follow-up for mid-size polar perf.

**Comment by eellison** [comment]
cc @hameerabbasi

**Review by blaine-rister — APPROVED** [review]
LGTM! Thanks for the thorough follow-up on review comments.

**Comment by eellison** [comment]
@hameerabbasi sorry, why doesnt this occur by default? potentially you answered this in another thread.

**Comment by amjames** [comment]
@eellison We have not enabled the complex wrapper by default as it needs some field testing, enabling the flag would lead to `polar` being lowered in much the same way. However the native tensor's real/imag components will be handled as two distinct contagious tensors and are only converted back at the graph boundary. This decomposition operates on two strided views and re-interleaves immediately. The two approaches are equivalent for a single op, in a real multi op graph the compile time wrapper would likely win.

I would recommend closing this, and instead using the flag to enable the whole graph transformation. 

`torch._functorch.config.enable_complex_wrapper = True`

**Comment by eellison** [comment]
Gotcha, makes sense. is there an issue for turning that on by default ? otherwise we are going to continue to get prs like this.

> The two approaches are equivalent for a single op, in a real multi op graph the compile time wrapper would likely win

I would just start by decomposing everything, and fusing, as much as possible, and then worry about interleaving. I'm not convinced theres a huge perf difference there anyway.

**Comment by simranbarnwal** [comment]
> Gotcha, makes sense. is there an issue for turning that on by default ? otherwise we are going to continue to get prs like this.
> 
> > The two approaches are equivalent for a single op, in a real multi op graph the compile time wrapper would likely win
> 
> I would just start by decomposing everything, and fusing, as much as possible, and then worry about interleaving. I'm not convinced theres a huge perf difference there anyway.

Does the wrapper support complex graph I/O? I was planning on something similar. To decompose everything, then eliding the real(complex(x)) and complex(real(x)) chains at op boundaries. 

The decomposition lets the complex survive in the FX graph!

**Comment by eellison** [comment]
I haven't really been following the recent work here, which was @hameerabbasi , @ezyang , and @amjames.

I would propose we just decompose everything. but, I'm fine with whatever has an actual path towards being on by default. See old hack a month i wrote up:





**torch.compile Support For Complex Numbers**

Background: there are two types of complex number representations:

**Interleaved**:

Tensor of Structs:  
\- \[(float, complex), (float, complex), ….\]

**Non-Interleaved**

Struct of Tensors  
\- \[float, float, float …  
\- \[complex, complex, complex, …\]

Pytorch Eager Uses an [Interleaved Representation](https://pytorch.org/docs/stable/complex_numbers.html#complex-numbers). However, this maps poorly to both cublas and triton. As Natalia pointed out [here](https://github.com/pytorch/pytorch/issues/125718#issuecomment-2106043924), there’s only tensorcore support for floating point numbers, not complex. Additionally, triton does not have native support. See discussion [here](https://github.com/pytorch/pytorch/issues/98161#issuecomment-1496426359). New API: triton split   
Spotty cublas coverage

**Proposal:**

We are going to decompose all complex operators inside of torch.compile to the non-interleaved format. For user-visible tensors, which escape the graph, we will cat the real and tensor to eager interleaved format.

The other implementation would be to support interleaved execution all the way through inductor. [https://github.com/pytorch/pytorch/issues/98161\#issuecomment-1494949775](https://github.com/pytorch/pytorch/issues/98161#issuecomment-1494949775)

Perf difference here seems to be negligible and it’s not worth the additional complexity in inductor. Simulating the difference between strided and non-strided access for a single operator (worst case behavior, no fusion) non-interleaved gets `.0707` while interleaved gets `.0705`, and both are faster than eager.

```py
import torch
import triton
torch.set_default_device("cuda")

@torch.compile()
def mul_decomposed(x, y, z):
    real_part = y * z
    imag_part = x * z
    
    # Stack in the correct order and along the last dimension
    stacked = torch.stack([real_part, imag_part], dim=-1)
    return stacked

x_comp = torch.rand([2048, 2048], dtype=torch.complex64)
x = x_comp.imag
y = x_comp.real
z = torch.rand([2048, 2048], dtype=torch.float32)

out_complex = torch.view_as_complex(mul_decomposed(x, y, z))
out_eager = x_comp * z

torch.testing.assert_close(out_complex, out_eager)
print("compiled perf", triton.testing.do_bench(lambda: mul_decomposed(x, y, z)))
print("eager perf", triton.testing.do_bench(lambda: x_comp * z))

@torch.compile()
def interleaved(x, y):
    return x * y

x = torch.rand([2048, 2048], dtype=torch.float64)
y = torch.rand([2048, 2048])
x * y
interleaved(x, y)

print("compiled perf", triton.testing.do_bench(lambda: interleaved(x, y)))
print("eager perf", triton.testing.do_bench(lambda: x * y))
```

Implementation Details:

1\. Register additional complex-only decompositions

See, this [PR](https://github.com/pytorch/pytorch/pull/110740/files#diff-c2b212e90b5c83183704f45942662dcc243a03fe64615a7acb95798efc077201R273-R282). We should check opinfos for operators which have native inductor lowerings but which we fallback for complex to help drive coverage.

2\. Make sure that successive complex constructors and decompositions get removed. Should also work with views, broadcasts.

```py
a = inp.view(complex_type)
real, imag = a.real, b.imag
```

See: https://github.com/pytorch/pytorch/pull/111773  
.  
3\. Make AOT Dispatcher guard, decompose semantics of conj. See [https://github.com/pytorch/pytorch/issues/145093](https://github.com/pytorch/pytorch/issues/145093).

`Conj` is a lazy negation of complex part of tensor. We should guard, and lower that out as operators in the graph.

**Comment by hameerabbasi** [comment]
> Does the wrapper support complex graph I/O? I was planning on something similar. To decompose everything, then eliding the real(complex(x)) and complex(real(x)) chains at op boundaries.

If by "complex graph I/O" you mean "receive arrays with `torch.complex*` dtypes, convert them, and convert the output back", the answer is yes. See https://github.com/pytorch/pytorch/pull/169832 and https://github.com/pytorch/pytorch/pull/167621, among other related PRs.

There are also [(nightly) docs](https://docs.pytorch.org/docs/main/user_guide/torch_compiler/torch.compiler_complex_number_support.html) on this.

**Comment by simranbarnwal** [comment]
> > Does the wrapper support complex graph I/O? I was planning on something similar. To decompose everything, then eliding the real(complex(x)) and complex(real(x)) chains at op boundaries.
> 
> If by "complex graph I/O" you mean "receive arrays with `torch.complex*` dtypes, convert them, and convert the output back", the answer is yes. See #169832 and #167621, among other related PRs.
> 
> There are also [(nightly) docs](https://docs.pytorch.org/docs/main/user_guide/torch_compiler/torch.compiler_complex_number_support.html) on this.

What is preventing us from turning the flag on by default? 

A problem I noticed was dead `aten.is_pinned` nodes in the graph when `enable_complex_wrapper` is set. 

```
  %is_pinned_default : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (), kwargs = {input: %clone_default, device: None})
```
`torch.polar(r, theta)` produces six of them.

`ComplexTensor.__new__` (`torch/_subclasses/complex_tensor/_core.py:56-88`) queries pinning three times per construction: once for the `pin_memory=` argument to `_make_wrapper_subclass`, and twice more for the `real.is_pinned() != imag.is_pinned()` assertion. `decompose_complex_in_graph` retraces the function under `make_fx`, and proxy-tensor tracing records every aten call including pure metadata queries. The resulting bool is consumed in Python, never by another node, so each call is left in the graph with `num_users=0`.

Is running a DCE at the end of `decompose_complex_in_graph` an appropriate solution for this. I can raise a PR. Alternatively, we can hoist the redundant `is_pinned()` calls in `__new__` so pinning is queried once.

  Repro:

  ```python
  import torch, torch._functorch.config

  def f(r, theta):
      return torch.polar(r, theta)

  with torch._functorch.config.patch(enable_complex_wrapper=True):
      torch.compile(f)(torch.rand(8, 24), torch.randn(8, 24))
  # inspect the post-AOT graph: dead aten.is_pinned nodes are present
  ```

**Comment by blaine-rister** [comment]
@eellison @hameerabbasi  thanks for the context. We weren't aware that a similar feature already existed. We can try reusing that instead of introducing a new decomposition for `polar`.

**Comment by hameerabbasi** [comment]
> Is running a DCE at the end of `decompose_complex_in_graph` an appropriate solution for this. I can raise a PR.

I'm happy to review.

> What is preventing us from turning the flag on by default?

Quoting [a previous comment](https://github.com/pytorch/pytorch/pull/192315#issuecomment-5259067305):

> We have not enabled the complex wrapper by default as it needs some field testing, enabling the flag would lead to `polar` being lowered in much the same way.

One such issue is the superfluous `is_pinned` nodes you discovered; and field testing is the best way to discover such issues.

**Comment by ezyang** [comment]
deferring to Hameer here

**Comment by hameerabbasi** [comment]
I feel it's best to focus efforts in one direction here; and given that significant work has gone into ensuring coverage and testing correctness of the (gated) subclass rewriting, we should work towards hardening it and turning it on by default. We can close this PR if that's okay for you @simranbarnwal?

**Comment by simranbarnwal** [comment]
> I feel it's best to focus efforts in one direction here; and given that significant work has gone into ensuring coverage and testing correctness of the (gated) subclass rewriting, we should work towards hardening it and turning it on by default. We can close this PR if that's okay for you @simranbarnwal?

It makes sense to focus on turning on the flag by default. I can close this PR.
